### PR TITLE
Feature/add presentation hook

### PR DESCRIPTION
**Pull request by narinluangrath**
I branched off of the current HEAD of `rewrite/add-deck`. I can't really demonstrate the functionality of the hook yet because of some bugs in `rewrite/add-deck`. Once those are resolved, I'll merge pull those changes into this PR and finish off the presentation code.

### CODE REVIEW

**Comment by kiraarghy** [comment]
Hey @narinluangrath I had a look at this today, with some help from @kitten we managed to get it working and it works pretty well, great work on getting it working with chromecast! 

However we have a suggestion to refactor some of the core state so that we can just send an object over to the presentation window with the state, so maybe something like:

```javascript
sendMessage({
  currentSlide: 1,
  currentSlideElement: 0
});
```

Then that filters through the cast component's state via context.

From my perspective I think that seems to be more of a deterministic patter and may be easier to implement/ maintain, interested in hearing your perspective on this? 😄

File: src/hooks/usePresentation.js [hunk 1]
```diff
@@ -3,25 +3,35 @@ import { useEffect, useState, useCallback, useMemo, useRef } from 'react';
...
-    if (!messageHandlers[key]) {
-      setMessageHandlers({ ...messageHandlers, [key]: handler });
-    }
-  }
+  // add message handler handles adding and removing handlers
+  const addMessageHandler = useCallback((handler, key) => {
+    const keyHandlers = (messageHandlers.current[key] =
+      messageHandlers.current[key] || []);
+    keyHandlers.push(handler);
+    return () => {
```
Comment (narinluangrath) [inline src/hooks/usePresentation.js]: I thought only `useEffect` accepts a cleanup function?

File: src/hooks/usePresentation.js [hunk 2]
```diff
@@ -3,25 +3,35 @@ import { useEffect, useState, useCallback, useMemo, useRef } from 'react';
...
   const [errors, setErrors] = useState([]);
 
-  const addMessageHandler = (handler, key) => {
-    if (!messageHandlers[key]) {
-      setMessageHandlers({ ...messageHandlers, [key]: handler });
-    }
-  }
+  // add message handler handles adding and removing handlers
+  const addMessageHandler = useCallback((handler, key) => {
+    const keyHandlers = (messageHandlers.current[key] =
```
Comment (narinluangrath) [inline src/hooks/usePresentation.js]: I didn't know you could do that, neat!

**Comment by narinluangrath** [comment]
@kiraarghy Yes! I like that a **lot** better. It will actually make communication easier for both ends, we don't need a `syncedDispatch`. I'm imagining something like

`Deck.js`
```
const [state, dispatch] = useDeck(...);
const { 
  sendMessage, 
  isConnected, 
  startConnection, 
  terminateConnection, 
  isReceiver, 
  addMessageHandler 
} = usePresentation();

// on page load, start the connection
useEffect(() => {
 startConnection();
 return terminateConnection; // cleanup
}, []);

// (controller) when the Deck state changes, trigger a `sendMessage` 
useEffect(() => {
  if (isConnected && !isReceiver) {
    sendMessage(state);
  }
}, [state, isConnected, isReceiver]);

// (receiver) update the Deck state when you receive the message
useEffect(() => {
  if (isConnected && isReceiver) {
    addMessageHandler(newDeckState => {
      // Use the `dispatch` function to update
      // the state. Ideally, add a type to the
      // reducer to accept a full state replacement
    })
  }
}, [isConnected, isReceiver])
```

If we have some state for `Slide.js` we're need to repeat the above code in that file as well.

**Comment by narinluangrath** [comment]
Feel free to merge

**Pull request merged**